### ISSUE TITLE
TMC2208 settings don't persist over PSU power cycle

### ISSUE THREAD

**Issue by swapsCAPS** [MarlinFirmware/Marlin#10723]
### Description

PSU power cycle clears TMC2208 settings

### Steps to Reproduce

1. Power off board and machine
2. Power on both
3. Send `M122`
4. msteps are set to 8 ( print head will happily ram into frame :p )
5. Send `M502`
6. msteps are set to 16
7. Send `M500`, settings stored, all good
8. Power off machine, leave board on 5V power
9. Send `M122`
10. msteps are set to 256 (and `0x00...` driver regs)
11. Power on machine
12. msteps are set to 8 again, regardless of previous `M500`

**Expected behavior:** Settings persist over power cycle

**Actual behavior:** Drivers get reset somehow

#### Additional Information

* Running recent bugfix-1.1.x on an MKS 1.4 Board with Watterott TMC2208s
* Using OctoPrint PSU control with a relais. And set: `#define POWER_SUPPLY 0`
* Files: https://github.com/swapsCAPS/Marlin/tree/ft-5-boegs

Honestly if the drivers need power for them to be able to persist settings I'm afraid the firmware can't really do much about it, when a user uses an external PSU switch.
Currently I've disabled `EEPROM_SETTINGS` and set OctoPrint to run `M502` before each print.
But this is not an elegant fix.

**Comment by teemuatlut** [comment]
Microsteps is not among the settings that is stored in EEPROM for TMC drivers. Nor is there a command to change them on the fly.
Microsteps reported as 256 is when the communication has failed and the response was 0x0UL, where msteps=0 means native 256.

There are some things we need to do better for handling main power when reading/writing to the drivers, but that's something I'll be doing for 2.0 release.

**Comment by swapsCAPS** [comment]
I understand. But how come they are set to 8 right after power-on. Is that just their default?
I've set the microstepping to 16 in the firmware. And an `M502` also sets that value.
But after a PSU power cycle the drivers are reset to 8 msteps.
I can repeat the process. Every PSU off/on they're set to 8 msteps and needs a `502` to set them to the correct value.
Is this expected behavior as the chips **lose** their settings at power loss?

**Comment by thinkyhead** [comment]
> I've set the microstepping to 16 in the firmware. And an M502 also sets that value.

I'm not sure why it's not getting your default setting (16x) on boot, as it does with `M502`, because `M501` (aka `settings.load`) also calls `reset_stepper_drivers` to set the defaults. And `settings.load` is called at boot time.

**Comment by swapsCAPS** [comment]
Does a PSU power cycle invoke `settings.load` though? I can imagine it only gets called when the board boots initially.

I think the following is happening:
- When I turn on the machine. The board gets power through USB. 
- Tries to load the settings.
- But the PSU is off, so writing over UART fails as the 2208s need power from the PSU.
- Switch on the PSU (via OctoPrint and a relay) and now the settings have not been set.

Is there a way for Marlin to know if the board started getting 12V? Maybe we could invoke `settings.load` after?

**Comment by teemuatlut** [comment]
Okey so here's how it's gonna work when I get to it after 1.1.9.

* On boot time Marlin will try to write the settings to the driver.
* If the drivers don't have high voltage power, the write will fail but the library will remember the settings
* When `M80` is invoked, Marlin will call `push` to sync the settings to the driver

I haven't yet decided if I'm going to force PSU on on boot time anyway.

If you control the PSU from OctoPrint or something else, you will need to send `M80` to Marlin to sync the settings! Even if the wire wouldn't be connected to the motherboard.

Concerning your problem, there really shouldn't be a situation where msteps would 8, if that is not what you configured in your hard coded settings.

**Comment by thinkyhead** [comment]
Ah, I see… I missed that it was a 12V PSU thing.

@teemuatlut — Does the `microsteps` method in the current TMC2130Stepper library already store the set value to the shadow registers?

The reason I ask is that when `AUTO_POWER_CONTROL` is enabled the `M80` handler calls `push` on all the TMC drivers, so all that would be needed is to enable `AUTO_POWER_CONTROL`.

That of course assumes that Octoprint is using `M80` to turn on the PSU.

**Comment by teemuatlut** [comment]
Everything (-ish) is stored. That's because when we want to set a bit in a register, we have to write the whole register again, just with a new 32bit pattern.

It's possible that the only issue here is that while `M80` does `push` (sync) to the registers, `AUTO_POWER_CONTROL` might not. It's one of those things where the dev doesn't use something so it easily gets overlooked or just forgotten.

**Comment by swapsCAPS** [comment]
I think using `M80` is a good approach.
It looks like the PSU control plugin on OctoPrint also supports sending gcode after PSU on/off. 
That's nice as you can then still use an external relay like I'm using.
The reason I am seeing 8 msteps is because `MS1` and `MS2` are not connected and the driver thinks it's in standalone mode I guess.
http://learn.watterott.com/silentstepstick/pinconfig/#tmc2208-step-configuration
For now I will set that newly found gcode setting to send `M502` after PSU on : )

**Comment by pmbroth** [comment]
Just a note, it may be muddying the water, but wanted to give some information.  On my MKS Gen 1.4 board, if I do not  Initialize the eprom 3 -4 times using the lcd controller, none of my firmware uploads or changes ever get stored.  I don't know if it is my board or not, but that is what I have learned to do every time, I want to make changes.

**Comment by pmbroth** [comment]
I'm running tmc2208 and 8825 drivers

**Comment by boelle** [comment]
@swapsCAPS is this still an issue if using bugfix 2.0?

**Comment by boelle** [comment]
@thinkyhead i think we can close this one

**Comment by EsQueue81** [comment]
The following info would have saved me quite some time when I found this thread. Since some of our boards are powered up at all times with 5v, when we send the M80 command to turn on our printers, it instantly sends the tries to initialize the stepper drivers. Since many 12v/24v supplies aren't full power instantly, the initialization fails. Loading the EEPROM settings for starting a print forces initialization. At the time of writing this, marlin added a #define PSU_POWERUP_DELAY function. Add a value in milliseconds after. In my case, I tried #define PSU_POWERUP_DELAY 1000 but 1 second would work most but not all the time due to my large PS and long start up times. I just  #define PSU_POWERUP_DELAY 2000. Two seconds works 100% of the time.
